src/tools/rbd/action/Export.cc

```diff
@@ -86,12 +365,159 @@ class C_Export : public Context
   SimpleThrottle &m_throttle;
   librbd::Image &m_image;
   bufferlist m_bufferlist;
+  uint64_t m_dest_offset;
   uint64_t m_offset;
   uint64_t m_length;
   int m_fd;
 };
 
-static int do_export(librbd::Image& image, const char *path, bool no_progress)
+static int do_export_v2(librbd::Image& image, librbd::image_info_t &info, int fd,
+		        uint64_t period, int max_concurrent_ops, utils::ProgressContext &pc)
+{
+  int r = 0;
+  size_t offset = 0;
+  // header
+  bufferlist bl;
+  bl.append(utils::RBD_IMAGE_BANNER_V2);
+
+  // size in header
+  uint64_t size = info.size;
+  ::encode(size, bl);
+
+  __u8 tag;
+  uint64_t length;
+  // encode order
+  tag = RBD_EXPORT_IMAGE_ORDER;
+  length = 8;
+  ::encode(tag, bl);
+  ::encode(length, bl);
+  ::encode(uint64_t(info.order), bl);
+
+  // encode format
+  tag = RBD_EXPORT_IMAGE_FORMAT;
+  uint8_t old_format;
+  r = image.old_format(&old_format);
+  if (r < 0) {
+    return r;
+  }
+  length = 8;
+  ::encode(tag, bl);
+  ::encode(length, bl);
+  if (old_format)
+    ::encode(uint64_t(1), bl);
+  else
+    ::encode(uint64_t(2), bl);
+
+  // encode features
+  tag = RBD_EXPORT_IMAGE_FEATURES;
+  uint64_t features;
+  image.features(&features);
+  length = 8;
+  ::encode(tag, bl);
+  ::encode(length, bl);
+  ::encode(features, bl);
+
+  // encode stripe_unit and stripe_count
+  tag = RBD_EXPORT_IMAGE_STRIPE_UNIT;
+  uint64_t stripe_unit;
+  stripe_unit = image.get_stripe_unit();
+  length = 8;
+  ::encode(tag, bl);
+  ::encode(length, bl);
+  ::encode(stripe_unit, bl);
+
+  tag = RBD_EXPORT_IMAGE_STRIPE_COUNT;
+  uint64_t stripe_count;
+  stripe_count = image.get_stripe_count();
+  length = 8;
+  ::encode(tag, bl);
+  ::encode(length, bl);
+  ::encode(stripe_count, bl);
+
+  // encode end tag
+  tag = RBD_EXPORT_IMAGE_END;
+  ::encode(tag, bl);
+
+  // write bl to fd.
+  r = bl.write_fd(fd);
+  if (r < 0) {
+    return r;
+  }
+  offset += bl.length();
+
+  lseek(fd, offset, SEEK_SET);
```

Comment:
Why is lseek needed? I suppose fd position is set correctly at this point? Also, we might write to stdout...